File: 백준/Silver/24060. 알고리즘 수업 － 병합 정렬 1/알고리즘 수업 － 병합 정렬 1.c
#include <stdio.h>

int arr[500000];
int tmp[500001];
int cnt = 0, result = -1, n, k;

void merge(int* A, int p, int q, int r);
void merge_sort(int* A, int p, int r);


int main(void) {
	scanf("%d %d", &n, &k);
	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
	}
	merge_sort(arr, 0, n - 1);
	printf("%d", result);
	
	return 0; 
}

void merge(int* A, int p, int q, int r) {
	int i = p, j = q + 1, t = 1;
	while (i <= q && j <= r) {
		if (A[i] <= A[j]) {
			tmp[t++] = A[i++];
		}
		else tmp[t++] = A[j++];
	}
	while (i <= q) tmp[t++] = A[i++];
	while (j <= r) tmp[t++] = A[j++];
	i = p, t = 1;
	while (i <= r) {
		A[i++] = tmp[t++];
		cnt++;
		if (cnt == k) {
			result = A[i - 1];
			break;
		}
	}
}

void merge_sort(int* A, int p, int r) {
	if (p < r) {
		int q = (p + r) / 2;
		merge_sort(A, p, q);
		merge_sort(A, q + 1, r);
		merge(A, p, q, r);
	}
}

File: 백준/Silver/24060. 알고리즘 수업 － 병합 정렬 1/알고리즘 수업 － 병합 정렬 1.py
import sys

cnt = 0
result = -1

def merge(arr, left, mid, right):
    global cnt, result
    sorted = []
    i = left
    j = mid + 1
    while(i <= mid and j <= right):
        if(arr[i] <= arr[j]):
            sorted.append(arr[i])
            i += 1
        else:
            sorted.append(arr[j])
            j += 1
    while(i <= mid):
        sorted.append(arr[i])
        i += 1
    while(j <= right):
        sorted.append(arr[j])
        j += 1
    l = left
    t = 0
    while(l <= right):
        arr[l] = sorted[t]
        cnt += 1
        if(cnt == k):
            result = arr[l]
            break
        l += 1
        t += 1
            

def merge_sort(arr, left, right):
    if(left < right):
        mid = (left + right) // 2
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)
        
            

n, k = map(int, sys.stdin.readline().split())
arr = list(map(int, sys.stdin.readline().split()[:n]))
merge_sort(arr, 0, n - 1)
print(result)